api/all/src/main/java/io/opentelemetry/api/common/AttributesBuilder.kt

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package io.opentelemetry.api.common
+
+import io.opentelemetry.api.common.AttributeKey.Companion.booleanArrayKey
+import io.opentelemetry.api.common.AttributeKey.Companion.booleanKey
+import io.opentelemetry.api.common.AttributeKey.Companion.doubleArrayKey
+import io.opentelemetry.api.common.AttributeKey.Companion.doubleKey
+import io.opentelemetry.api.common.AttributeKey.Companion.longArrayKey
+import io.opentelemetry.api.common.AttributeKey.Companion.longKey
+import io.opentelemetry.api.common.AttributeKey.Companion.stringArrayKey
+import io.opentelemetry.api.common.AttributeKey.Companion.stringKey
+import java.util.*
+import java.util.function.Predicate
+
+/** A builder of [Attributes] supporting an arbitrary number of key-value pairs.  */
+interface AttributesBuilder {
```

Comment:
As a general rule instead of builders I think we should offer a Kotlin DSL. E.g.

```
mySpan.setAttributes {
    addAttribute("my_key", "my"value")
    addAttribute("my_key", "my"value")
}
```

Do we want this to be the only way to interact with our APIs, or should we offer alternatives?

Reply:
FWIW the OTel spec only requires setting attributes. I haven't seen anything about introspecting added values, although I can see how that would be useful for an implementer & a consumer https://opentelemetry.io/docs/specs/otel/trace/api/#set-attributes

Reply:
Consensus: DSL first, add extras if needed after testing usability